### DISCUSSION TITLE
`BindParameter` does not infer the type of `UserDefinedType` instances

### DISCUSSION THREAD

**Discussion by buck-ross** [sqlalchemy/sqlalchemy#9754] [Usage Questions] [answered]
## Problem

I think I may have found a bug with the way `UserDefinedType` instances are handled by `BindParameter` which leads to unexpected behavior when passing literal values to database functions.

In short, `test_func = func.DUMMY_FUNC(FooType('b'))` where `FooType` is a subclass of `UserDefinedType` does not correctly interpret the type of the bound parameter.
Running `str(list(test_func.params().clauses))` yields the output `[BindParameter('%(XXXXXXXXXXXXXXX DUMMY_FUNC)s', FooType('b'), type_=NullType())]` where the `type_` parameter is erroneously set to `NullType()` instead of `FooType()`.

## Full Example

Consider a fresh PostgreSQL database initialized as follows:

```sql
-- Create the table:
CREATE TABLE test_table (
        rowid INTEGER,
        foo VARCHAR,
        PRIMARY KEY(rowid)
);

-- Define some dummy function that prepends "DUMMY: " to its input:
CREATE FUNCTION dummy_func(val IN VARCHAR) RETURNS VARCHAR 
        LANGUAGE plpgsql
AS $function$
        BEGIN
                RETURN 'DUMMY: ' || val;
        END;
$function$

-- Populate the table with some sample data:
INSERT INTO test_table(rowid, foo) VALUES (1, 'FOO: a'), (2, 'FOO: b');
```

From here, we can create a simple custom datatype `FooType` as follows:

```python
from sqlalchemy.types import UserDefinedType

class FooType(UserDefinedType):
        cache_ok = True

        def __init__(self, value=None):
                self.value = value

        def __repr__(self):
                return f'{self.__class__.__name__}({self.value!r})'

        # Prepend "FOO: " to the data before insertion / update:
        def bind_processor(self, dialect):
                def processor(instance):
                        return f'FOO: {instance.value}'
                return processor

        # Remove the leading "FOO: " on retrieval from the database:
        def result_processor(self, dialect, coltype):
                def processor(value):
                        return self.__class__(value[5:])
                return processor
```

We can use this type in a custom table like so:

```python
from sqlalchemy import Column, Integer
from sqlalchemy.orm import declarative_base

class MyTable(declarative_base()):
        __tablename__ = 'test_table'
        rowid = Column(Integer, primary_key=True)
        foo = Column(FooType) # Use the custom type for the "foo" column of the table

        def __repr__(self):
                return f'{self.__class__.__name__}(rowid={self.rowid!r}, foo={self.foo!r})'
```

And finally attempt to run a simple test case:

```python
from sqlalchemy import create_engine, func
from sqlalchemy.orm import sessionmaker

# Connect to the local PostgreSQL database:
engine = create_engine(
        'postgresql+psycopg2://postgres:*****@localhost:5432/postgres'
)
SessionMaker = sessionmaker()
SessionMaker.configure(bind=engine)

with SessionMaker() as session:
        # Build a simple query that involves passing a literal value to a database function:
        func_literal = func.DUMMY_FUNC(FooType('b'))
        query = session.query(MyTable).filter(
                func.DUMMY_FUNC(MyTable.foo) == func_literal
        )

        # Inspect the query itself:
        print('QUERY: ', str(query).replace('\n', ' '))
        print('PARAMS: ', list(func_literal.params().clauses))

        # Try to run the query (success is expected, but it fails anyway):
        try:
                res = query.all()
                print(f'SUCCESS: {res}')
        except BaseException as err:
                print(f'FAILURE: {err}')
```

When we run this test, we would expect to see the following:

> QUERY:  SELECT test_table.rowid AS test_table_rowid, test_table.foo AS test_table_foo  FROM test_table  WHERE DUMMY_FUNC(test_table.foo) = DUMMY_FUNC(%(DUMMY_FUNC_1)s)
PARAMS:  [BindParameter('%(XXXXXXXXXXXXXXX DUMMY_FUNC)s', FooType('b'), type_=FooType(None))]
SUCCESS: [MyTable(rowid=2, foo=FooType('b'))]

But instead, the query fails, and we get the following output:

> QUERY:  SELECT test_table.rowid AS test_table_rowid, test_table.foo AS test_table_foo  FROM test_table  WHERE DUMMY_FUNC(test_table.foo) = DUMMY_FUNC(%(DUMMY_FUNC_1)s)
PARAMS:  [BindParameter('%(XXXXXXXXXXXXXXX DUMMY_FUNC)s', FooType('b'), type_=NullType())]
FAILURE: (psycopg2.ProgrammingError) can't adapt type 'FooType'
[SQL: SELECT test_table.rowid AS test_table_rowid, test_table.foo AS test_table_foo 
FROM test_table 
WHERE DUMMY_FUNC(test_table.foo) = DUMMY_FUNC(%(DUMMY_FUNC_1)s)]
[parameters: {'DUMMY_FUNC_1': FooType('b')}]
(Background on this error at: https://sqlalche.me/e/20/f405)

## Workaround (Hack)

It seems possible to implement a work around for this issue by customising the `TypeEngine` used during the type inference process.
To do this, we need to define a custom type engine class like so:

```python
from sqlalchemy.types import TypeEngine

class HackedTypeEngine(TypeEngine):
        def _resolve_for_literal(self, value):
                # Honor UserDefinedType instances, but fall back to normal type inferences:
                if isinstance(value, UserDefinedType):
                        return value.__class__()
                return super()._resolve_for_literal(value)
```

And then we need to modify `FooType` to use the new type engine like so:

```python
class FooType(UserDefinedType):
        __sa_type_engine__ = HackedTypeEngine()

        ... # Insert the rest of FooType's code here
```

Although the above modifications will allow the provided test case to succeed, this solution seems like it may have some unintended side-effects that I'm not fully aware of, so use it at your own risk.

Bottom line: I'm not entirely sure if this is a bug or simply a use-case that wasn't fully considered. Either way, it seems like it merits further discussion.

## Relevant Code Sections

- [`BindParameter`'s `type_` inference proceedure inside `__init__`](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/sql/elements.py#L2041-L2071)
- [`_resolve_value_to_type` - where the type inference itself seems to take place](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/sql/sqltypes.py#L3802-L3826)

**Comment by CaselIT** [comment]
Hi,

That's the expected behaviour. SQLAchemy types define SQL types, not python ones, so doing the following is wrong.
> In short, test_func = func.DUMMY_FUNC(FooType('b')) where FooType is a subclass of UserDefinedType does not correctly interpret the type of the bound parameter.

Like you don't do `func.some_function(Integer(42))` you do not do the same with user defined types

Maybe you are looking for Cast? or function should have a `type_` kwarg

**Comment by buck-ross** [comment]
@CaselIT Ok, that makes sense, but there is still one more problem.

I've refactored the code of `FooType` to split into a python type `Foo` and a SQL type `FooCol`. Please note that the actual implementation of the `Foo` class used by my application is much more complex, and **cannot be replaced by `str`** or another primitive type.

```python
from sqlalchemy.types import UserDefinedType

class Foo:
        def __init__(self, value=None):
                self.value = value

        def __repr__(self):
                return f'{self.__class__.__name__}({self.value!r})'


class FooCol(UserDefinedType):
        cache_ok = True

        def bind_processor(self, dialect):
                def processor(instance):
                        return f'FOO: {instance.value}'
                return processor

        def result_processor(self, dialect, coltype):
                def processor(value):
                        return Foo(value[5:])
                return processor
```

The function invocation can now be accomplished as follows:

```python
func_literal = func.DUMMY_FUNC(type_coerce(Foo('b'), FooCol))
query = session.query(MyTable).filter(
        func.DUMMY_FUNC(MyTable.foo) == func_literal
)
```

This works, but the problem is that it requires manual type coercion each time I want to use a `Foo` instance as a bound parameter, which forces me to use statements like `type_coerce(val, FooCol) if isinstance(val, Foo) else val` whenever I want to bind a parameter that may take on more than one type.

This problem is smaller, but is still a big determent to the overall readability and usability of my application. Is there any way to either

1. tell SQLAlcmemy that `Foo` should be coerced to `FooCol` by default, the same way `int` is coerced to `INTEGER` and `str` to `VARCHAR` or
2. define a `GenericFunction` for `func.DUMMY_FUNC` that will automatically coerce its parameters into `FooCol`?

Thank you very much for your help so far! If you are able to help me resolve this last remaining issue, I will gladly mark the question as answered.

**Reply by CaselIT** [reply, marked as answer]
> tell SQLAlcmemy that Foo should be coerced to FooCol by default, the same way int is coerced to INTEGER and str to VARCHAR or

no there is nothing public that allows augmenting the internal mapping.

> define a GenericFunction for func.DUMMY_FUNC that will automatically coerce its parameters into FooCol?

This is the currently supported option. Just implement a class with the name you want to use (it's case sensitive) and import it at least once so that's registered.
You can take a look at array_agg that does something like what you want: https://github.com/sqlalchemy/sqlalchemy/blob/cef8adac8ae48c5c6c36138ed0b59067c1ef78ed/lib/sqlalchemy/sql/functions.py#L1575-L1614

**Reply by buck-ross** [reply]
@CaselIT That's perfect! Thanks again!
